Additional events

noobanidus opened the pull request:
As discussed, the following new events:

```
AnimalTameEvent
BlockFarmlandTrampleEvent
BlockPlaceEvent
CriticalHitEvent
CropGrowPostEvent
CropGrowPreEvent
EnchantmentLevelSetEvent
EntityMountEvent
EntityTravelToDimensionEvent
ExplosionDetonateEvent
ExplosionStartEvent
ItemFishedEvent
LivingDestroyBlockEvent
LivingExperienceDropEvent
LivingKnockBackEvent
LootingLevelEvent
MinecartCollisionEvent
MinecartInteractEvent
MobGriefingEvent
PlayerCloseContainerEvent
PlayerItemPickupEvent
PlayerLeftClickBlockEvent
PlayerRightClickEmptyEvent
PlayerRightClickItemEvent
PlayerVisibilityEvent
PotionBrewPostEvent
PotionBrewPreEvent
ProjectileImpactArrowEvent
ProjectileImpactFireballEvent
ProjectileImpactThrowableEvent
SleepingLocationCheckEvent
SleepingTimeCheckEvent
```

[Here's the test script (rename from .txt) to I've used to test all of the events.](https://github.com/CraftTweaker/CraftTweaker/files/4333423/test.zs.txt)

Documentation forthcoming.

noobanidus committed:
First in additional events: AnimalTameEvent.

noobanidus committed:
Implement BlockEvent.FarmlandTrampleEvent.

noobanidus committed:
Critical Hit event.

noobanidus committed:
EnchantmentLevelSet event.

noobanidus committed:
EntityMountEvent.

noobanidus committed:
Explosion events.

noobanidus committed:
ItemFishedEvent.

noobanidus committed:
CropGrowth events.

noobanidus committed:
BlockPlace event.

noobanidus committed:
MobGriefing.

noobanidus committed:
EntityTravelToDimensionEvent.

noobanidus committed:
ItemTooltipEvent.

noobanidus committed:
LivingDestroyBlockEvent.

noobanidus committed:
LivingExperienceDropEvent

noobanidus committed:
LivingKnockBackEvent.

noobanidus committed:
LootingLevelEvent.

noobanidus committed:
Remove ItemTooltipEvent?

noobanidus committed:
Minecart events.

noobanidus committed:
PlayerContainerClose event.

This, along with Open, are not (or no longer) cancelable.

noobanidus committed:
PlayerItemPickupEvent

noobanidus committed:
Player Visibility Event.

noobanidus committed:
PlayerLeftClickBlockEvent.

noobanidus committed:
Oops.

noobanidus committed:
RightClickEmptyEvent, not every PlayerInteractEvent is cancellable.

noobanidus committed:
PlayerRightClickItemEvent.

noobanidus committed:
SaplingGrowTreeEvent.

noobanidus committed:
Sleeping Checks.

noobanidus committed:
PotionBrewEvents.

noobanidus committed:
ProjectileImpactEvents.

noobanidus committed:
Whoops.

noobanidus committed:
Fix CropGrowEvent names.

noobanidus committed:
Some minor tweaks.

noobanidus committed:
Fix some minor issues.

noobanidus committed:
Add hand and ItemStack handlers for PlayerInteractEvent.

noobanidus committed:
Screw the Terrain Generation bus.

noobanidus committed:
Final exorcism of things.

noobanidus committed:
It's already a getter.

noobanidus committed:
EntityMount is cancelable; limit EnchantmentLevelSet to 0-30.

noobanidus committed:
This event is also cancellable.

kindlich commented on CraftTweaker2-API/src/main/java/crafttweaker/api/event/MTEventManager.java:
```diff
@@ -100,6 +99,37 @@ public void clear() {
+        elPotionBrewPost.clear();
+        elProjectileImpactArrow.clear();
+        elProjectileImpactFireball.clear();
+        elProjectileImpactThrowable.clear();
```
While you're at it, could you check if the other events also properly add their clears here?

I'm not sure if I did it for every event I added there ^^'

noobanidus replied:
I'll write a script to do it, I think it's probably safer that way than relying on my faulty memory.

noobanidus replied:
Turns out that two of mine were missing along with 4 others. I've added them now.

kindlich commented on CraftTweaker2-API/src/main/java/crafttweaker/api/event/PlayerItemPickupEvent.java:
```diff
@@ -0,0 +1,18 @@
+
+@ZenClass("crafttweaker.event.PlayerItemPickupEvent")
+@ZenRegister
+public interface PlayerItemPickupEvent extends IPlayerEvent {
```
How does this event differ from `PlayerPickupItemEvent `
<https://github.com/CraftTweaker/CraftTweaker/blob/1.12/CraftTweaker2-API/src/main/java/crafttweaker/api/event/PlayerPickupItemEvent.java>

noobanidus replied:
It differs by using a different name but the same event, I didn't catch it. Whoops.

noobanidus replied:
Scratch that, I was wrong: `MCPlayerPickupItemEvent` is an implementation of `EntityItemPickupEvent`. That event is fired whenever a player runs into an EntityItem.

`PlayerItemPickupEvent` is an implementation of the `ItemPickupEvent` of the `PlayerEvent` class and fires when the item has been placed into the inventory.

kindlich commented on CraftTweaker2-API/src/main/java/crafttweaker/api/event/PlayerRightClickBlockEvent.java:
```diff
@@ -4,7 +4,7 @@
 import stanhebben.zenscript.annotations.*;
 
-@ZenClass("crafttweaker.event.PlayerInteractBlockEvent")
+@ZenClass("crafttweaker.event.PlayerRightClickBlockEvent")
```
I'm not sure if I want this change, since it would break existing scripts that use the current event's name.

Alternative would be to create a marker interface named PlayerInteractBlockEvent that extends the PlayerRightClickBlock event, and have the MCPlayerInteractBlockEvent implement both.
That way it would not make a difference whether they used one name or the other in their event handler function.

noobanidus replied:
This is a completely separate event with a completely different function; I'm not actually renaming anything.

kindlich commented on CraftTweaker2-API/src/main/java/crafttweaker/api/event/PlayerVisibilityEvent.java:
```diff
@@ -0,0 +1,16 @@
+@ZenRegister
+public interface PlayerVisibilityEvent extends IPlayerEvent {
+  @ZenMethod
+  void modifyVisibility (double modifier);
```
Why is one a getter and the other one a method?

noobanidus replied:
The default Event implementation is currently:

```java
        public void modifyVisibility(double mod)
        {
            visibilityModifier *= mod;
        }
```

As it's specifically a modifier *of the value of the modifier*, I felt it would be more confusing to have it as a getter/setter combo -- as you would set the value to something, then get the value only to get a different value.

kindlich commented on CraftTweaker2-API/src/main/java/crafttweaker/api/event/ProjectileImpactArrowEvent.java:
```diff
@@ -0,0 +1,48 @@
+  void setDamage (double damage);
+
+  @ZenSetter("knockbackStrength")
+  void setKnockbackStrength (int knockbackStrength);
```
Just to clarify:
Do they also have a way to get that value, or only a way to set it?

noobanidus replied:
These methods are short-hand accessors into the actual EntityArrow to avoid writing an interface for EntityArrow. As far as I'm aware, `knockbackStrength` for EntityArrow is a private int, and it only has a setter; I decided to not implement a getter, but didn't want to short-change people on things that they would otherwise be able to access.

kindlich commented on CraftTweaker2-API/src/main/java/crafttweaker/api/event/ProjectileImpactArrowEvent.java:
```diff
@@ -0,0 +1,48 @@
+  void setIsCritical(boolean isCritical);
+
+  @ZenGetter("pickupStatus")
+  String getPickupStatus();
```
For IEventHasResult, you used three methods to check the different status' and three methods to set them.
Here you use one string to get them, and three methods to set them.

noobanidus replied:
Perhaps IEventHasResult it could do with a default String implementation that returns the type instead of relying on 3 variables -- one of the things I wasn't sure about with the implementation was to how to specifically handle getting it.

noobanidus replied:
I've pushed a change to make IEventHasResult match this and other methods.

kindlich commented on CraftTweaker2-MC1120-Main/src/main/java/crafttweaker/mc1120/events/CommonEventHandler.java:
```diff
@@ -33,13 +31,19 @@
+import net.minecraftforge.event.entity.minecart.MinecartInteractEvent;
 import net.minecraftforge.event.entity.player.*;
 import net.minecraftforge.event.furnace.FurnaceFuelBurnTimeEvent;
+import net.minecraftforge.event.terraingen.SaplingGrowTreeEvent;
```
Since you're not using the event, could you remove the unused import ^^

noobanidus replied:
Sure.

kindlich commented on CraftTweaker2-MC1120-Main/src/main/java/crafttweaker/mc1120/events/handling/MCSaplingGrowTreeEvent.java:
```diff
@@ -0,0 +1,55 @@
+/*package crafttweaker.mc1120.events.handling;
```
You removed the interface but only commented out the implementation?

noobanidus replied:
I meant to delete this.

kindlich reviewed (changes requested):
Your indentation for most files differs from ours:
We usually use 4 whitespaces, you are using 2.


Requesting changes due to the renaming of PlayerInteractBlockEvent's ZenClass.
@jaredlll08 What do you think about that?

noobanidus commented:
Indentation is my fault, I'll tweak that.

I'm pretty sure that was just my mistake in using the PlayerInteractEvent as a template and forgetting to rename it in the class. PlayerInteractEvent still exists as normal.

noobanidus committed:
Fix indents.

noobanidus committed:
More indentations fixes.

noobanidus committed:
Fix rename failure.

noobanidus commented:
The rename was a failure on my part, I misunderstood.

noobanidus committed:
Unify allow/deny/default into a single string result as per other systems.

noobanidus committed:
Add missing `clear` calls.

noobanidus committed:
Remove RightClickEmpty, it only fires on the client side.

Various other tweaks.

noobanidus committed:
Allow access to the cancellation result.

jaredlll08 commented:
Is there anything wrong with this now?

noobanidus committed:
Unify some class/interface names.

noobanidus committed:
Rename ProjectileImpactEvent to IProjectileImpactEvent.

noobanidus commented:
I'm happy that I've addressed all of the issues that Kindlich raised. I've renamed some of the "base" events into Interfaces as that seems to match how everything else is handled. I have tested every event works, that those that can be cancelled are cancelled, etc, but I haven't test everything single piece of the puzzle, specifically accessor to Forge event parameters.

I'm hopeful that if I'm just providing access to a setter/getter then there should be no issues resulting from that, as it should Just Work.
